File: cpp/src/leet_code/easy/has_cycle.cpp
// cpp/src/leet_code/common/has_cycle.cpp

#include "has_cycle.hpp"
#include <iostream>
#include <unordered_set>

using namespace std;

bool Solution::hasCycle(ListNode *head)
{
    auto node = head;
    unordered_set<ListNode *> visited;
    while (node != NULL)
    {
        if (visited.find(node) != visited.end())
            return true;
        visited.insert(node);
        node = node->next;
    }
    return false;
}

bool Solution::hasCycle2(ListNode *head)
{
    auto fast = head;
    auto slow = head;
    while (fast != NULL && fast->next != NULL)
    {
        fast = fast->next->next;
        slow = slow->next;
        if (fast == slow)
            return true;
    }
    return false;
}

File: python/src/dsa/linked_list/has_cycle.py
# python/src/dsa/linked_list/has_cycle.py

# 141. Linked List Cycle
# https://leetcode.com/problems/linked-list-cycle/description/

from typing import Optional

from dsa.linked_list.linked_list import ListNode


class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        visited = set()
        node = head
        while node:
            if node in visited:
                return True
            visited.add(node)
            node = node.next

        return False

    def hasCycle2(self, head: Optional[ListNode]) -> bool:
        fast, slow = head, head
        while fast and fast.next:
            fast, slow = fast.next.next, slow.next
            if slow == fast:
                return True

        return False
